File: frontend/web/src/app/core/services/sector.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, catchError, map, of } from 'rxjs';

import { environment } from '../../../environments/environment';
import { HoldSectorRequest, HoldSectorResponse, Sector } from '../models/sector.models';
import { PagedResult, toPublishStatus, toTicketingMode } from '../models/catalog.models';

/** Same ordinal-vs-name normalization CatalogService does — see coerceEnum. */
function normalizeSector(raw: Sector): Sector {
  return {
    ...raw,
    ticketingMode: toTicketingMode(raw.ticketingMode),
    status: toPublishStatus(raw.status),
  };
}

@Injectable({ providedIn: 'root' })
export class SectorService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = environment.apiBaseUrl;

  /**
   * GET /api/sectors?productId= — Published only, public.
   *
   * `date` (`yyyy-MM-dd`) is what makes `remainingCapacity` meaningful for a DailyEntry product:
   * that mode counts capacity per `(sector, date)`, so the backend returns null rather than guess
   * when no date is supplied. Ignored by every other mode. Callers showing a date picker should
   * re-request on each change, since the availability answer changes with it.
   */
  getSectors(productId: string, date?: string | null): Observable<PagedResult<Sector>> {
    const params: Record<string, string> = { productId, pageSize: '100' };
    if (date) params['date'] = date;

    return this.http
      .get<PagedResult<Sector>>(`${this.baseUrl}/sectors`, { params })
      .pipe(map((result) => ({ ...result, items: result.items.map(normalizeSector) })));
  }

  /** POST /api/sectors/{id}/hold — Redis atomic hold, TTL 5 min. */
  hold(sectorId: string, request: HoldSectorRequest): Observable<HoldSectorResponse> {
    return this.http.post<HoldSectorResponse>(`${this.baseUrl}/sectors/${sectorId}/hold`, request);
  }

  /** POST /api/sectors/holds/{holdId}/release — best-effort early release of a still-live hold
   * (e.g. the buyer switched to a different spot/date, or a sibling hold in the same cart failed
   * to purchase). Always resolves, even on a network/server error — callers only ever want to
   * give capacity back on a best-effort basis, never to block on it or surface a failure toast for
   * a release the buyer didn't directly ask for. */
  release(holdId: string): Observable<void> {
    return this.http
      .post<void>(`${this.baseUrl}/sectors/holds/${holdId}/release`, {})
      .pipe(catchError(() => of(void 0)));
  }
}

File: frontend/mobile/lib/services/sector_service.dart
import 'package:dio/dio.dart';

import '../core/api_client.dart';
import '../models/api_error.dart';
import '../models/requests/hold_sector_request.dart';
import '../models/responses/hold_sector_response.dart';
import '../models/responses/paged_result.dart';
import '../models/responses/sector_response.dart';
import 'api_exception.dart';

/// Sector browsing + the Redis atomic hold — mirrors `SectorService` in
/// `frontend/web`.
class SectorService {
  bool _isSuccess(int? code) => code != null && code >= 200 && code < 300;

  Never _handleError(Response response) {
    ApiError apiError;
    try {
      final body = response.data;
      apiError = body is Map<String, dynamic> ? ApiError.fromJson(body) : ApiError(message: body?.toString());
    } catch (_) {
      apiError = ApiError(message: response.data?.toString());
    }
    throw ApiException(statusCode: response.statusCode ?? 0, apiError: apiError);
  }

  /// GET /api/sectors?productId= — Published only, public.
  ///
  /// [date] is only meaningful for a DailyEntry product: capacity there is counted per
  /// (Sector, date), so without it the backend leaves `remainingCapacity` null and the caller
  /// cannot tell a full day from an empty one. Every other mode ignores it.
  Future<PagedResult<SectorResponse>> getSectors(String productId, {DateTime? date}) async {
    final response = await apiClient.get('Sectors', queryParameters: {
      'productId': productId,
      'pageSize': 100,
      if (date != null) 'date': _formatDate(date),
    });
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return PagedResult.fromJson(response.data as Map<String, dynamic>, SectorResponse.fromJson);
  }

  /// `DateOnly` on the wire — the backend binds yyyy-MM-dd, and sending a full ISO timestamp
  /// would drag the device's timezone into a calendar-date decision.
  static String _formatDate(DateTime date) =>
      '${date.year.toString().padLeft(4, '0')}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';

  /// POST /api/sectors/{id}/hold — Redis atomic decrement/lock, TTL 5 min.
  /// Requires auth — callers should route to LoginScreen on a 401.
  Future<HoldSectorResponse> hold(String sectorId, HoldSectorRequest request) async {
    final response = await apiClient.post('Sectors/$sectorId/hold', data: request.toJson());
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return HoldSectorResponse.fromJson(response.data as Map<String, dynamic>);
  }

  /// POST /api/sectors/holds/{holdId}/release — best-effort early release of a still-live hold
  /// (e.g. the buyer switched to a different spot/date, or a sibling hold in the same cart failed
  /// to purchase). Deliberately swallows every failure — callers only ever want to give capacity
  /// back on a best-effort basis, never to block on it or surface an error for a release the buyer
  /// didn't directly ask for.
  Future<void> release(String holdId) async {
    try {
      await apiClient.post('Sectors/holds/$holdId/release');
    } catch (_) {
      // Best-effort — an already-expired holdId, or a transient network error, is fine to ignore.
    }
  }
}
